esrally/telemetry.py

```diff
@@ -1408,6 +1409,199 @@ def record(self):
+
+        self.metrics_store.put_value_cluster_level(
+            "ingest_pipeline_cluster_time", self.ingest_pipeline_cluster_time, "ms", meta_data=metadata
+        )
+        self.metrics_store.put_value_cluster_level(
+            "ingest_pipeline_cluster_failed", self.ingest_pipeline_cluster_failed, meta_data=metadata
+        )
+
+    def _record_node_level_pipeline_stats(self, stats, cluster_name, node_name):
+        # Node level statistics are calculated per-benchmark execution. Stats are collected at the beginning, and end of
+        # each benchmark
+        metadata = {"cluster_name": cluster_name}
+        ingest_pipeline_node_count = stats.get("count", 0) - self.start_stats[cluster_name][node_name]["total"].get("count", 0)
+        ingest_pipeline_node_time = stats.get("time_in_millis", 0) - self.start_stats[cluster_name][node_name]["total"].get(
+            "time_in_millis", 0
+        )
+        ingest_pipeline_node_failed = stats.get("failed", 0) - self.start_stats[cluster_name][node_name]["total"].get("failed", 0)
+
+        self.ingest_pipeline_cluster_count += ingest_pipeline_node_count
+        self.ingest_pipeline_cluster_time += ingest_pipeline_node_time
+        self.ingest_pipeline_cluster_failed += ingest_pipeline_node_failed
+
+        self.metrics_store.put_value_node_level(node_name, "ingest_pipeline_node_count", ingest_pipeline_node_count, meta_data=metadata)
+        self.metrics_store.put_value_node_level(node_name, "ingest_pipeline_node_time", ingest_pipeline_node_time, "ms", meta_data=metadata)
+        self.metrics_store.put_value_node_level(node_name, "ingest_pipeline_node_failed", ingest_pipeline_node_failed, meta_data=metadata)
+
+    def _record_pipeline_level_processor_stats(self, pipeline, pipeline_name, cluster_name, node_name):
+        for processor_name, processor_stats in pipeline.items():
+            start_stats_processors = self.start_stats[cluster_name][node_name]["pipelines"][pipeline_name]
+
+            if processor_name not in start_stats_processors:
+                self.logger.warning(
+                    "Cannot determine Ingest Pipeline stats in %s for %s (processor was not defined at the start of the benchmark).",
+                    pipeline_name,
+                    processor_name,
+                )
+                continue
+            # We have an individual processor obj, which contains the stats for each individual processor
+            if processor_name != "total":
+
+                metadata = {
+                    "processor_name": processor_name,
+                    "type": processor_stats.get("type", None),
+                    "ingest_pipeline": pipeline_name,
+                    "cluster_name": cluster_name,
+                }
+
+                ingest_pipeline_processor_count = processor_stats.get("stats", {}).get("count", 0) - start_stats_processors[
+                    processor_name
+                ].get("stats", {}).get("count", 0)
```

Comment:
Can we use two variables here? The current formatting is really confusing to me. (Same below)